src/zon.zig

```diff
@@ -0,0 +1,661 @@
+
+        buf.clearRetainingCapacity();
+        try tree.renderError(note, buf.writer(gpa));
+        try self.sema.pt.zcu.errNote(
+            .{
+                .base_node_inst = try ip.trackZir(gpa, .main, .{
+                    .file = self.file_index,
+                    .inst = .main_struct_inst,
+                }),
+                .offset = .{ .token_abs = note.token + @intFromBool(note.token_is_prev) },
+            },
+            err_msg,
+            "{s}",
+            .{buf.items},
+        );
+    }
+
+    try self.sema.pt.zcu.failed_files.putNoClobber(gpa, self.file, err_msg);
+    return error.AnalysisFail;
+}
+
+const Ident = struct {
+    bytes: []const u8,
+    owned: bool,
+
+    fn deinit(self: *Ident, allocator: Allocator) void {
+        if (self.owned) {
+            allocator.free(self.bytes);
+        }
+        self.* = undefined;
+    }
+};
+
+fn ident(self: LowerZon, token: Ast.TokenIndex) !Ident {
+    var bytes = self.file.tree.tokenSlice(token);
+
+    if (bytes[0] == '@' and bytes[1] == '"') {
+        const gpa = self.sema.gpa;
+
+        const raw_string = bytes[1..bytes.len];
+        var parsed = std.ArrayListUnmanaged(u8){};
+        defer parsed.deinit(gpa);
+
+        switch (try std.zig.string_literal.parseWrite(parsed.writer(gpa), raw_string)) {
+            .success => {
+                if (std.mem.indexOfScalar(u8, parsed.items, 0) != null) {
+                    return self.fail(.{ .token_abs = token }, "identifier cannot contain null bytes", .{});
+                }
+                return .{
+                    .bytes = try parsed.toOwnedSlice(gpa),
+                    .owned = true,
+                };
+            },
+            .failure => |err| {
+                const offset = self.file.tree.tokens.items(.start)[token];
+                return self.fail(
+                    .{ .byte_abs = offset + @as(u32, @intCast(err.offset())) },
+                    "{}",
+                    .{err.fmtWithSource(raw_string)},
+                );
+            },
+        }
+    }
+
+    return .{
+        .bytes = bytes,
+        .owned = false,
+    };
+}
+
+fn identAsNullTerminatedString(self: LowerZon, token: Ast.TokenIndex) !NullTerminatedString {
+    var parsed = try self.ident(token);
+    defer parsed.deinit(self.sema.gpa);
+    const ip = &self.sema.pt.zcu.intern_pool;
+    return ip.getOrPutString(self.sema.gpa, self.sema.pt.tid, parsed.bytes, .no_embedded_nulls);
+}
+
+const FieldTypes = union(enum) {
+    st: struct {
+        ty: Type,
+        loaded: InternPool.LoadedStructType,
+    },
+    un: struct {
+        ty: Type,
+        loaded: InternPool.LoadedEnumType,
+    },
+    none,
+
+    fn init(ty: ?Type, sema: *Sema) !@This() {
+        const t = ty orelse return .none;
+        const ip = &sema.pt.zcu.intern_pool;
+        switch (t.zigTypeTagOrPoison(sema.pt.zcu) catch return .none) {
+            .Struct => {
+                try t.resolveFully(sema.pt);
```

Comment:
Why resolve these types fully? ZON shouldn't care about type layout, only the field types (and I suppose defaults).